src/renderers/shared/fiber/ReactFiberHostContext.js

```diff
@@ -34,61 +34,69 @@ module.exports = function<T, P, I, TI, C, CX>(
   } = config;
 
-  type ContainerState = {
-    rootInstance : C,
-    contextFibers : Array<Fiber | null> | null,
-    contextValues : Array<CX | null> | null,
-    contextDepth: number
-  };
-
-  function createContainerState(rootInstance) {
-    return {
-      rootInstance,
-      contextFibers: null,
-      contextValues: null,
-      contextDepth: -1,
-    };
-  }
-
-  // If we meet any portals, we'll go deeper.
-  let containerStack : Array<ContainerState> = [];
-  let containerDepth : number = -1;
+  // Context stack is reused across the subtrees.
+  // We use a null sentinel on the fiber stack to separate them.
+  let contextFibers : Array<Fiber | null> | null = null;
+  let contextValues : Array<CX | null> | null = null;
+  let contextDepth : number = -1;
+  // Current root instance.
+  let rootInstance : C | null = null;
+  // A stack of outer root instances if we're in a portal.
+  let portalStack : Array<C | null> = [];
+  let portalDepth : number = -1;
 
   function getRootHostContainer() : C {
-    if (containerDepth === -1) {
-      throw new Error('Expected to find a root container.');
+    if (rootInstance == null) {
+      throw new Error('Expected root container to exist.');
     }
-    const containerState = containerStack[containerDepth];
-    const {rootInstance} = containerState;
     return rootInstance;
   }
 
-  function pushHostContainer(portalHostContainer) {
-    containerDepth++;
-    if (containerDepth === containerStack.length) {
-      containerStack[containerDepth] = createContainerState(portalHostContainer);
+  function pushHostContainer(nextRootInstance : C) {
+    if (rootInstance == null) {
+      // We're entering a root.
+      rootInstance = nextRootInstance;
     } else {
-      const containerState = containerStack[containerDepth];
-      containerState.rootInstance = portalHostContainer;
-      containerState.contextFibers = null;
-      containerState.contextValues = null;
-      containerState.contextDepth = -1;
+      // We're entering a portal.
+      // Save the current root to the portal stack.
+      portalDepth++;
+      portalStack[portalDepth] = rootInstance;
+      rootInstance = nextRootInstance;
+      // Delimit subtree context with a sentinel so we know where to pop later.
+      if (contextFibers != null && contextValues != null) {
+        contextDepth++;
+        contextFibers[contextDepth] = null;
+        contextValues[contextDepth] = null;
+      }
     }
   }
 
   function popHostContainer() {
-    if (containerDepth === -1) {
-      throw new Error('Already reached the root.');
+    if (portalDepth === -1) {
+      // We're popping the root.
+      rootInstance = null;
+      contextDepth = -1;
+    } else {
+      // We're popping a portal.
+      // Restore the root instance.
+      rootInstance = portalStack[portalDepth];
+      portalStack[portalDepth] = null;
+      portalDepth--;
+      // If we pushed any context while in a portal, we need to roll it back.
```

Comment:
How can you ever pop a host container with a context deeper than one `null`? Wouldn't this always be exactly one `null`?

Because any new context provider will be below the portal in the tree and pop on the way up to the portal.

Think of this problem in terms of a normal execution stack. You wouldn't need multiple arrays and these special cases for unwinding in a normal execution stack other than when you unwind for errors so you shouldn't need it here.